File: go/circular-buffer/circular_buffer.go
// Package circular implements simple routines for circular bufffer
package circular

import "errors"

// Buffer holds items and current read index
type Buffer struct {
	data      []byte
	count     int
	readIndex int
}

// NewBuffer creates and returns a new Buffer
func NewBuffer(size int) *Buffer {
	return &Buffer{
		data: make([]byte, size),
	}
}

// ReadByte returns byte from current read postion
func (b *Buffer) ReadByte() (byte, error) {
	if b.count == 0 {
		return 0, errors.New("cannot read empty buffer")
	}

	ri := b.readIndex % len(b.data)
	v := b.data[ri]
	b.readIndex = ri + 1
	b.count--

	return v, nil
}

// WriteByte appends byte to buffer returns error if buffer is full
func (b *Buffer) WriteByte(v byte) error {
	if b.count == len(b.data) {
		return errors.New("cannot write to full buffer")
	}

	wi := (b.readIndex + b.count) % len(b.data)
	b.data[wi] = v
	b.count++

	return nil
}

// Overwrite writes value to buffer, if full overwrites the first value
func (b *Buffer) Overwrite(v byte) {
	if b.count == len(b.data) {
		b.ReadByte()
	}
	b.WriteByte(v)
}

// Reset resets buffer
func (b *Buffer) Reset() {
	b.count = 0
}

File: solutions/csharp/circular-buffer/2/CircularBuffer.cs
using System;

public class CircularBuffer<T>
{
    private readonly T[] buffer;
    private int count = 0;
    private int readIndex = 0;

    public CircularBuffer(int capacity)
    {
        buffer = new T[capacity];
        count = 0;
        readIndex = 0;
    }

    public T Read()
    {
        if (count == 0)
        {
            throw new InvalidOperationException("Cannot read empty buffer.");
        }

        readIndex = readIndex % buffer.Length;
        var value = buffer[readIndex++];
        count--;

        return value;
    }

    public void Write(T value)
    {
        if (count == buffer.Length)
        {
            throw new InvalidOperationException("Cannot write to full buffer.");
        }
        
        var writeIndex = (readIndex + count) % buffer.Length;
        buffer[writeIndex] = value;
        count++;
    }

    public void Overwrite(T value)
    {
        if (count < buffer.Length)
        {
            Write(value);
        }
        else
        {
            buffer[readIndex] = value;
            readIndex++;
        }
    }

    public void Clear()
    {
        count = 0;
    }
}